unit_tests/ex_suffix_expander.cc

```diff
@@ -9,13 +9,177 @@
+    } else if (unit == TB) {
+      return static_cast<uintmax_t>(1) << 40;
+    } else if (unit == PB) {
+      return static_cast<uintmax_t>(1) << 50;
+    } else {
+      // TODO: error handling
+      return static_cast<uintmax_t>(0);
+    }
+  }
+
+  static inline TextView canonicalize(TextView unit) {
+    if (iequal(unit, "b") || iequal(unit, "byte")) {
+      return B;
+    } else if (iequal(unit, "k") || iequal(unit, "kb")) {
+      return KB;
+    } else if (iequal(unit, "m") || iequal(unit, "mb")) {
+      return MB;
+    } else if (iequal(unit, "g") || iequal(unit, "gb")) {
+      return GB;
+    } else if (iequal(unit, "t") || iequal(unit, "tb")) {
+      return TB;
+    } else if (iequal(unit, "p") || iequal(unit, "pb")) {
+      return PB;
+    } else {
+      // TODO: error handling
+      return B;
+    }
+  }
+};
+
+struct Duration {
+  static constexpr TextView SECOND = "second";
+  static constexpr TextView MINUTE = "minute";
+  static constexpr TextView HOUR   = "hour";
+  static constexpr TextView DAY    = "day";
+  static constexpr TextView WEEK   = "week";
+
+  static constexpr TextView baseUnit = "second";
+
+  static uintmax_t convertToBase(TextView const unit) {
+    if (unit == SECOND) {
+      return static_cast<uintmax_t>(1);
+    } else if (unit == MINUTE) {
+      return static_cast<uintmax_t>(60);
+    } else if (unit == HOUR) {
+      return static_cast<uintmax_t>(3'600);
+    } else if (unit == DAY) {
+      return static_cast<uintmax_t>(86'400);
+    } else if (unit == WEEK) {
+      return static_cast<uintmax_t>(604'800);
+    } else {
+      // TODO: error handling
+      return 0;
+    }
+  }
+  static inline TextView canonicalize(TextView unit) {
+    if (iequal(unit, "s") || iequal(unit, "sec") || iequal(unit, "second")) {
+      return SECOND;
+    } else if (iequal(unit, "m") || iequal(unit, "min") || iequal(unit, "minute")) {
+      return MINUTE;
+    } else if (iequal(unit, "h") || iequal(unit, "hour")) {
+      return HOUR;
+    } else if (iequal(unit, "d") || iequal(unit, "day")) {
+      return DAY;
+    } else if (iequal(unit, "w") || iequal(unit, "week")) {
+      return WEEK;
+    } else {
+      // TODO: error handling
+      return SECOND;
+    }
+  }
+};
+
+} // namespace metric
+
+// (potentially) modifies the pass-in view. returns the unit
+template <typename Metric> inline TextView rExtractUnit(TextView &src) noexcept {
+  src.rtrim_if(&isspace);
+  if (isdigit(src.back())) {
+    // no unit specification, defaulting to the smallest unit
+    return Metric::baseUnit;
+  } else {
+    // take suffix from right to left -- from the first non-space char
+    // (inclusive) to the first digit (exclusive)
+    // can't call split/take_suffix_if because we don't want to discard any char
+    // can't call suffix_if because it can't deal with the edge case that
+    // [0, pos] is a multiplier (string)
+    auto const pos  = src.rfind_if([](char const c) { return isdigit(c) || c == ' '; });
```

Comment:
Unfortunate `suffix_if` can't be used. Might do something like this though -
```
TextView unit = src;
src.rtrim_if([](char c){ return isdigit(c) || isspace(c); });
unit.remove_prefix(src.size());
```

Reply:
yea I meant to bring this up, basically there are two things that I think could make my code here more elegant:
1. a modify-return affix API (prolly only for the predicate-based *_if ones) that's not discarding the match. I understand that the current API is useful for string patterns that we have an axis/delimiter like in `aaa,bbb`, where we certainly don't care about the `,`. But what if the axis is missing? like what if we need to get the `bbb` suffix out of `aaabbb`, and on the next step go for `aaa`.
2. in addition to 1. if no match is found, move the whole src string to output. This is like treating the beginning/ending of the original string as "hard delimiters" that match anything.

I'm not sure if implementing 1/2 would have certain ramifications for TextView internally, just thought having them would be nice.

Reply:
I ended up writing `take_token_if` to do this. It would probably be a useful addition to `TextView`. Almost everything in the class now was added this way, by needing it for a specific parsing style.